### PR TITLE
Support arbitrary selectable text in Text component

### LINKED ISSUE #754: Hit test is broken on Text screen in RNTester app
### Environment

1. `react-native -v`: I'm on this commit:
```
commit 5a3ee83efeec124b3aedbcccc90f2a878b75917e (HEAD -> master, origin/master, origin/HEAD)
Author: Saad Najmi <saadnajmi2@gmail.com>
Date:   Wed Mar 31 23:18:16 2021 -0500
```
3. `npm ls react-native-macos`: `(empty)`
4. `node -v`: `v12.14.1`
5. `npm -v`: `6.13.4`
6. `yarn --version`: `1.21.1`
7. `xcodebuild -version`:
```
Xcode 12.4
Build version 12D4e
```

### Issue
When I run RNTester on Mac and go to Text examples, if I right click on the selectable text example, it will only show the correct context menu if I click within the top few pixels of the component.

### Steps to Reproduce

1. Clone repo
2. Run RNTester Mac
3. Go to Text example
4. Scroll to selectable text
5. Click and hold to select the text
6. Right click on the lower half of the text

### Expected Behavior

Context menu with copy should show.

### Actual Behavior

RN Context menu shows.

### Reproducible Demo

The first click in this video is on top half (works fine), second click is bottom half (bug).

https://user-images.githubusercontent.com/359157/113638221-36146b00-962b-11eb-8a26-48e8c10ad18e.mov

### Additional context

The bug is due to this line: https://github.com/microsoft/react-native-macos/blob/master/RNTester/js/components/RNTesterPage.js#L72
If I delete it, everything works fine. It seems that setting a `paddingTop` on `ScrollView` breaks hit tests.

### PR DESCRIPTION
**Pull request by shwanton**
#### Please select one of the following
- [ ] I am removing an existing difference between facebook/react-native and microsoft/react-native-macos :thumbsup:
- [ ] I am cherry-picking a change from Facebook's react-native into microsoft/react-native-macos :thumbsup:
- [x] I am making a fix / change for the macOS implementation of react-native
- [ ] I am making a change required for Microsoft usage of react-native

## Summary

Summary:

RN Mac's implementation of the `selectable` prop on `Text` only allows selecting the entire Text component and right click to copy. This diff makes the `Text.selectable` prop on Mac allow arbitrary selection. To do this we used `NSTextView` to render the `Text` component instead of RN Mac's custom text rendering, because it has a `selectable` property which gives us the behavior we want.

Differential Revision: https://phabricator.intern.facebook.com/D27484533

Tasks: T83817888

## Changelog

[macOS] [Fixed] - Support arbitrary selectable text in Text component
[macOS] [Fixed] - Add `focusable` and  `enableFocusRing` props to Text to control focus behavior

## Test Plan

before (RN iOS behavior)
CleanShot 2022-08-10 at 09 47 33

after

https://user-images.githubusercontent.com/96719/192903795-e2830a01-bff6-4e33-9099-8ea3535f733a.mp4

### CODE REVIEW

File: Libraries/Text/Text/RCTTextView.h
Comment (christophpurrer) [inline Libraries/Text/Text/RCTTextView.h:20]: This is the 'old' tracking number.
Please change to
// TODO(macOS GH#774)

Since this is a multi-line change, also add the beginning ```[``` and ending ```]```

File: packages/rn-tester/js/components/RNTesterPage.js
Comment (christophpurrer) [inline packages/rn-tester/js/components/RNTesterPage.js:70]: I think this is not needed anymore?

**Comment by lyahdav** [comment]
@shwanton seems like you can remove this line from the PR summary now?
> Note the change to RNTesterPage.js was required to fix https://github.com/microsoft/react-native-macos/issues/754

**Comment by Saadnajmi** [comment]
We probably want to do _a lot more_ testing for this specific change, as it's a fundamental difference for how we render text. Curious, have y'all done any performance impact testing with this changee?

**Comment by lyahdav** [comment]
We did not do any performance impact testing. Does this repo (internally to MSFT or in OSS) have any performance testing setup?

**Comment by Saadnajmi** [comment]
> We did not do any performance impact testing. Does this repo (internally to MSFT or in OSS) have any performance testing setup?

Not that I'm aware of. I guess the more generic question: What kind of testing was done with making sure that Text still rendered the same everywhere? I can imagine just running RN-Tester before and after this change is a good start (we'll probably do that too).


Also curious if you all heard back from RN Core about why they didn't use UITextView with iOS?

**Comment by lyahdav** [comment]
> What kind of testing was done with making sure that Text still rendered the same everywhere?

We went through all the Text examples in RN Tester and confirmed the before/after were identical.

> Also curious if you all heard back from RN Core about why they didn't use UITextView with iOS?

I'll let @shwanton answer this, he has an internal thread on this.

**Comment by christophpurrer** [comment]
> What kind of testing was done with making sure that Text still rendered the same everywhere?

Another data point -> we shipped Messenger Desktop (macOS) with that change end of 2021 and so far haven't been aware of any text rendering issues.

**Comment by shwanton** [comment]
> Also curious if you all heard back from RN Core about why they didn't use UITextView with iOS?

This Textview refactor happened in late 2017 so not many folks have full context on why we removed UITextView. However, it does look like the performance reasons should be now mitigated. One main callout was that "it does not instantiate UIView for virtual components which reduces memory utilization", which should be solved with Fabric.

RN Core recommended we run an experiment before switching out UITextView on iOS. This work is planned but has not been started yet.

We have been using this on Messenger Desktop all year & haven't seen any regressions or issues w/ NSTextView.
I can do some more general smoke tests & post the before/after shots in the test plan as well.

**Comment by shwanton** [comment]
When we added the change, we did a visual diff of the Text component in RN Tester. (The differences are highlighted in pink)
image
* The font smoothing isn't something we need
image
* The examples with images are different because they load random images
* The pink background on "With size and background color" isn't a difference, the background color is pink in code
image

**Comment by shwanton** [comment]
@Saadnajmi was there any other data you needed to help this PR get merged?

**Comment by Saadnajmi** [comment]
@Shwanton Sorry, I totally missed the screenshots you sent. Thanks! @mischreiber FYI

**Comment by lyahdav** [comment]
I have some concerns with this PR as-is:

1. It seems this brings in an older implementation of our internal changes. We have fixed some bugs, for example with the `hitTest:` method. @shwanton an example diff is D37429403. You might want to audit if there are any other changes we missed here.
2. There's a known issue internally where sometimes mouse clicks are lost in our RN Mac fork. We still haven't gotten to the bottom of it, but since that `hitTest` and `mouseDown` code are kind of hacky, I wonder if this is related. @shwanton are we confident the lost mouse clicks is related to our RCTTouchHandler changes and not this?
3. There are potentially other bugs with this implementation. @shwanton T124433678 might require changes to this code. Is that a task you could take on? If so I could do some knowledge transfer.

File: Libraries/Text/Text/RCTTextView.m
Comment (lyahdav) [inline Libraries/Text/Text/RCTTextView.m:215]: should we merge this with previous `#if TARGET_OS_OSX`?

File: Libraries/Text/Text/RCTTextView.m
Comment (lyahdav) [inline Libraries/Text/Text/RCTTextView.m:261]: This might be easy to miss when trying to understand this code. Basically the rest of this method is only used on iOS but it might not be obvious. @Saadnajmi I'm curious if you'd be ok with this sneaky `return` as-is.

**Comment by lyahdav** [comment]
@shwanton another example of a change we might want to make this to PR: D28557126

File: Libraries/Text/Text/RCTTextView.m
Comment (Saadnajmi) [reply Libraries/Text/Text/RCTTextView.m:261]: I noticed it. I haven't given too much thought to the actual code here quite yet, as I was prioritizing some of the other changes before this one :) 

_Technically_ , early returns are against our internal style guide ([aka.ms/objc-guide](https://microsoft.github.io/objc-guide/MethodsAndImplementations/EarlyReturns.html), but React Native macOS is kind of a special case. We generally prefer keeping diffs minimal and maybe this is the cleanest way to do so. I'll give it some thought :)

Comment (lyahdav) [reply Libraries/Text/Text/RCTTextView.m:261]: Interesting. I don't mind early returns in general because they act like Swift guard statements, but perhaps the branching here would be cleaner if at the top of `setTextStorage:` we had:
```
#if TARGET_OS_OSX
  [self setMacTextStorage:...];
  return;
#endif
```

Comment (Saadnajmi) [reply Libraries/Text/Text/RCTTextView.m:261]: Yeah, Swift guards are the case we _do_ have early returns :). Too bad there isn't Swift code in RN yet...

Comment (lyahdav) [reply Libraries/Text/Text/RCTTextView.m:261]: Another option:
```
#if TARGET_OS_OSX
- (void)setTextStorage:(NSTextStorage *)textStorage
{
...
}
#else // iOS
- (void)setTextStorage:(NSTextStorage *)textStorage
{
...
}
```

File: Libraries/Text/Text/RCTTextView.m
Comment (christophpurrer) [reply Libraries/Text/Text/RCTTextView.m:215]: We can't.
This is actually a bug as we learned this week > D39356602

File: Libraries/Text/Text/RCTTextView.m
Comment (Saadnajmi) [reply Libraries/Text/Text/RCTTextView.m:261]: Ah.. this is a fun one because it also supports Catalyst in the implementation (line 267)..

It sounds like refactoring could cause unintended bugs (like https://github.com/microsoft/react-native-macos/pull/1346#discussion_r967386554 ?) so perhaps we keep it in the safe state? I think I'm ok with the early return so long as we document with a comment why we can skip the rest? If the issue was font smoothing, I think it's worth noting we do have our own fontsmoothing implementation here: https://github.com/microsoft/react-native-macos/pull/329 )

File: Libraries/Text/Text/RCTTextView.m
Comment (lyahdav) [inline Libraries/Text/Text/RCTTextView.m:99]: @shwanton I see our fork still has focusRingMaskBounds. Should we keep it?

File: Libraries/Text/Text/RCTTextView.m
Comment (lyahdav) [inline Libraries/Text/Text/RCTTextView.m:189]: typo `desktop`

File: Libraries/Text/Text/RCTTextView.m
Comment (lyahdav) [inline Libraries/Text/Text/RCTTextView.m:472]: `RCTTextView`?

File: Libraries/Text/Text/RCTTextView.m
Comment (lyahdav) [inline Libraries/Text/Text/RCTTextView.m:473]: In our fork we override resignFirstResponder and canBecomeKeyView on the NSTextView subclass, not on RCTTextView. Do we need to move these there? Also we always return NO from canBecomeKeyView.

File: Libraries/Text/Text/RCTTextView.m
Comment (shwanton) [reply Libraries/Text/Text/RCTTextView.m:99]: Added this back in b08aae919d36fa87b03f1ee633e6e8148b66d8dc

File: Libraries/Text/Text/RCTTextView.m
Comment (lyahdav) [inline Libraries/Text/Text/RCTTextView.m:370]: @shwanton Our fork doesn't have drawFocusRingMask (just focusRingMaskBounds), is it needed?

File: Libraries/Text/Text/RCTTextView.m
Comment (shwanton) [inline Libraries/Text/Text/RCTTextView.m:375]: In our fork we removed the focus ring for selectable text since text was always un-focusable & never a key view (cc. @lyahdav). 

I believe the the behavior we want is when the TextView is `selectable`:
  1. it becomes a key view
  2. Is selectable with keyboard 
  3. Has a focus ring when focused

NOTE: I've added an extra text input to this example to verify key focus behavior.
CleanShot 2022-09-15 at 16 43 55

File: Libraries/Text/Text/RCTTextView.m
Comment (shwanton) [reply Libraries/Text/Text/RCTTextView.m:493]: Good call. RCTTextView can be the key view but I do think we want the actual NSTextView to be selected here.
After investigating, I'm not sure the `resignFirstResponder` fix is doing anything here.

See comment under focus view for why we should use `_selectable` to determine `canBecomeKeyView` value.

File: Libraries/Text/Text/RCTTextView.m
Comment (shwanton) [reply Libraries/Text/Text/RCTTextView.m:370]: see comment on `enableFocusRing` for proposed behavior

File: Libraries/Text/Text/RCTTextView.m
Comment (shwanton) [reply Libraries/Text/Text/RCTTextView.m]: fixed in b72d4bcaf5e482feeb784494cb2ab069a0aa26fe

File: Libraries/Text/Text/RCTTextView.m
Comment (shwanton) [reply Libraries/Text/Text/RCTTextView.m]: Fixed in b72d4bcaf5e482feeb784494cb2ab069a0aa26fe

File: Libraries/Text/Text/RCTTextView.m
Comment (lyahdav) [reply Libraries/Text/Text/RCTTextView.m:375]: @Saadnajmi thoughts on this? The default NSTextView behavior for selectable non-editable text is that it is in the key view loop, but tabbing to it doesn't show a focus ring, which is not ideal. In our fork we subclass NSTextView and make canBecomeKeyView return NO always. It's still accessible. IMO it doesn't make sense to have it be a tab stop. This would also put it in line with NSTextField's behavior when it's selectable and non-editable. Alternatively one option would be to expose a focusable property on Text for controlling this.

Comment (Saadnajmi) [reply Libraries/Text/Text/RCTTextView.m:375]: We have at least one client team that _wants_ Text to be a tab stop so they can focus on it, then do Cmd+C to copy text. So I think we should at least support `focusable`. IIRC, `selectable` effectively made it focusable in the old behavior. The "new" behavior having the two props handle separate things makes sense. 
@Ppatboyd FYI, since he worked on similar stuff on the windows (well, win32) side

Comment (lyahdav) [reply Libraries/Text/Text/RCTTextView.m:375]: I'm pretty sure even if it is a tab stop, when you're focused on an NSTextView and press Cmd+C it won't copy the text because it's not selected yet. I think you need to first press Cmd+A to select all, then Cmd+C. @shwanton perhaps you can verify?

Comment (shwanton) [reply Libraries/Text/Text/RCTTextView.m:375]: Correct, you will still need to select the text before it can be copied. It would be possible to automatically select the content when tabbed, but that is not expected behavior.

I'l add support for the `focusable` prop that will allow tab stops on TextViews.

**Comment by lyahdav** [comment]
Just so it's documented, we hit a bug recently related to this code where anytime you click on a virtual text node it will trigger a remount because `isHighlighted` is changing (but not used), which can lead to race condition bugs/crashes. @shwanton once we land D39742593 let's update this PR with that change.

File: Libraries/Text/Text/RCTTextView.m
Comment (shwanton) [inline Libraries/Text/Text/RCTTextView.m:133]: Keep previous behavior where `selectable={true}` will be `focusable` unless disabled.

File: Libraries/Text/Text/RCTTextView.m
Comment (shwanton) [inline Libraries/Text/Text/RCTTextView.m:506]: Use `focusable` prop to set first responder. if `_selectable` is true, `focusable` will be true

File: Libraries/Text/Text.js
Comment (shwanton) [inline Libraries/Text/Text.js:45]: The `supressHighlighting` does not work on macOS w/ NSTextView.
This was also causing a re-render of the Text component which lead to race conditions/crashes when tapping href links.

Comment (shwanton) [reply Libraries/Text/Text.js:45]: After investigating, highlighting does work on desktop w/o NSTextView. This does however cause the aforementioned crash when opening a native link.

We could either remove highlighting from desktop or try and come up with a version that works w/ NSTextView that does not have the same re-rendering behavior that caused the crashes.

Comment (lyahdav) [reply Libraries/Text/Text.js:45]: What does highlighting do on desktop? Is this different than selectable text? I believe on iOS the highlight behavior is that when you click the Text component it highlights it visually, but it highlights the entire component, without arbitrary selectable text (which is what we need).

Comment (shwanton) [reply Libraries/Text/Text.js:45]: Text Highlighting on iOS is only working w/ Paper (Fabric support is coming).

The entire text component will highlight when tapped. This is a mobile centric UX, similar to what happens when a tableview row is tapped. 

https://user-images.githubusercontent.com/96719/195733831-11fd5204-49ec-4f91-900b-7823cc234227.mp4

On Desktop, this might be implemented as a hover effect instead of the highlight on press.
For these reasons, it makes sense to remove highlighting from RN desktop on MacOS.

@Saadnajmi Let us know if this all makes sense.

**Review by mischreiber — APPROVED** [review]
Gotta fix the failing checks first, of course.

File: Libraries/Text/Text.js
Comment (Saadnajmi) [reply Libraries/Text/Text.js:45]: Did we already have any existing code trying to make 'suppressHighlighting' work on macOS? I'm happy to leave it as an iOS only prop.

Comment (shwanton) [reply Libraries/Text/Text.js:45]: `suppressHighlighting` actually worked w/o any customization for macOS. This is because it was using the custom text render from iOS implementation instead on NSTextView. However, this highlight should be handled by `hover` state instead of on `press` which is a very mobile centric UX.

**Comment by shwanton** [comment]
@mischreiber Fixed!

File: Libraries/Text/Text.js
Comment (Saadnajmi) [reply Libraries/Text/Text.js:45]: I see. While it's "technically" a regression, it sounds like it wasn't really something that should have worked on macOS to begin with. I'm happy to just leave it as is. @mischreiber fyi

Comment (mischreiber) [reply Libraries/Text/Text.js:45]: Yep, seems like a reasonable compromise for the time being.

**Comment by christophpurrer** [comment]
nice work!

**Pull request merged**
